### PR TITLE
Remove crate_type and crate_name attributes from compiler crates

### LINKED ISSUE #41701: Move attributes from crate source into Cargo.toml where appropriate
The compiler and std crates in-tree have a few attributes in their source code (lib.rs) that are conventionally set in the Cargo.toml file. I'd say this just wasn't cleaned up when rustbuild became the only build system, and should be cleaned up.

Possibly incomplete list of such attributes:

- `#![crate_name]`
- `#![crate_type]`

Note that some crates have different crate types in between Cargo.toml than in lib.rs (e.g., compiler crates seem to have only dylib in Cargo.toml, dylib and rlib in the source code). I'm not sure which one takes precedence, but my build directory seems to contain only dylibs and not rlibs for those crates, so I assume the Cargo.toml wins.

cc @alexcrichton -- is the above correct?

### PR DESCRIPTION
**Pull request by daniellockyer**
Fixes #41701.

### CODE REVIEW

**Comment by aidanhs** [comment]
Hi @neosilky, thanks for the PR! We'll check back now and again to make sure @brson or another reviewer gets to this soon.

**Comment by Mark-Simulacrum** [comment]
@brson Checking in, could you review this PR?

cc @alexcrichton

**Comment by brson** [comment]
Travis failure is spurious.

File: src/rtstartup/rsend.rs [hunk 1]
```diff
@@ -11,7 +11,6 @@
 // See rsbegin.rs for details.
 
 #![feature(no_core, lang_items, optin_builtin_traits)]
-#![crate_type="rlib"]
```
Comment (Mark-Simulacrum) [inline src/rtstartup/rsend.rs]: This needs to be re-added, as there is no associated `Cargo.toml`.

File: src/rtstartup/rsbegin.rs [hunk 2]
```diff
@@ -23,7 +23,6 @@
 // of other runtime components (registered via yet another special image section).
 
 #![feature(no_core, lang_items, optin_builtin_traits)]
-#![crate_type="rlib"]
```
Comment (Mark-Simulacrum) [inline src/rtstartup/rsbegin.rs]: This needs to be re-added, as there is no associated `Cargo.toml`.

**Review by Mark-Simulacrum — DISMISSED** [review]

**Comment by Mark-Simulacrum** [comment]
@neosilky Found a few more problems, but I think those should be the last things.

**Comment by neosilky** [comment]
@Mark-Simulacrum Ah yes! I've re-added in the latest commit.

**Comment by Mark-Simulacrum** [comment]
@brson I think this is ready for another review. Could you take a look?

**Comment by carols10cents** [comment]
Looks like there's a test failure?

```
[00:50:31] failures:
[00:50:31] 
[00:50:31] ---- [run-make] run-make/alloc-extern-crates stdout ----
[00:50:31] 	
[00:50:31] error: make failed
[00:50:31] status: exit code: 2
[00:50:31] command: "make"
[00:50:31] stdout:
[00:50:31] ------------------------------------------
[00:50:31] make[1]: Entering directory '/checkout/src/test/run-make/alloc-extern-crates'
[00:50:31] LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/alloc-extern-crates.stage2-x86_64-unknown-linux-gnu:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib:" '/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc' --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/alloc-extern-crates.stage2-x86_64-unknown-linux-gnu -L /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/alloc-extern-crates.stage2-x86_64-unknown-linux-gnu  fakealloc.rs
[00:50:31] LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/alloc-extern-crates.stage2-x86_64-unknown-linux-gnu:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib:" '/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc' --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/alloc-extern-crates.stage2-x86_64-unknown-linux-gnu -L /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/alloc-extern-crates.stage2-x86_64-unknown-linux-gnu  ../../../liballoc/lib.rs --cfg feature=\"external_crate\" --extern external=/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/alloc-extern-crates.stage2-x86_64-unknown-linux-gnu/libfakealloc.rlib
[00:50:31] thread 'main' panicked at 'Some tests failed', /checkout/src/tools/compiletest/src/main.rs:315
[00:50:31] Makefile:4: recipe for target 'all' failed
[00:50:31] make[1]: Leaving directory '/checkout/src/test/run-make/alloc-extern-crates'
[00:50:31] 
[00:50:31] ------------------------------------------
[00:50:31] stderr:
[00:50:31] ------------------------------------------
[00:50:31] make[1]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
[00:50:31] error: language item required, but not found: `panic_fmt`
[00:50:31] 
[00:50:31] error: language item required, but not found: `eh_personality`
[00:50:31] 
[00:50:31] error: aborting due to 2 previous errors
[00:50:31] 
[00:50:31] make[1]: *** [all] Error 101
[00:50:31] 
[00:50:31] ------------------------------------------
[00:50:31] 
[00:50:31] thread '[run-make] run-make/alloc-extern-crates' panicked at 'explicit panic', /checkout/src/tools/compiletest/src/runtest.rs:2467
[00:50:31] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[00:50:31] 
[00:50:31] 
[00:50:31] failures:
[00:50:31]     [run-make] run-make/alloc-extern-crates
[00:50:31] 
[00:50:31] test result: FAILED. 149 passed; 1 failed; 0 ignored; 0 measured
```

**Comment by arielb1** [comment]
Nice PR, @neosilky! There are however some test failures. Do you need help fixing them?

**Comment by arielb1** [comment]
compilation failure:
```
[00:03:07] error[E0432]: unresolved import `Substitution`
[00:03:07]   --> /checkout/src/librustc_errors/diagnostic.rs:12:5
[00:03:07]    |
[00:03:07] 12 | use Substitution;
[00:03:07]    |     ^^^^^^^^^^^^ no `Substitution` in the root
[00:03:07] 
[00:03:07] error[E0560]: struct `CodeSuggestion` has no field named `substitution_parts`
[00:03:07]    --> /checkout/src/librustc_errors/diagnostic.rs:209:13
[00:03:07]     |
[00:03:07] 209 |             substitution_parts: vec![Substitution {
[00:03:07]     |             ^^^^^^^^^^^^^^^^^^^ field does not exist - did you mean `substitutes`?
[00:03:07] 
[00:03:07] error[E0560]: struct `CodeSuggestion` has no field named `substitution_parts`
[00:03:07]    --> /checkout/src/librustc_errors/diagnostic.rs:220:13
[00:03:07]     |
[00:03:07] 220 |             substitution_parts: vec![Substitution {
[00:03:07]     |             ^^^^^^^^^^^^^^^^^^^ field does not exist - did you mean `substitutes`?
[00:03:07] 
[00:03:07] error: no field `substitution_parts` on type `&CodeSuggestion`
[00:03:07]   --> /checkout/src/librustc_errors/emitter.rs:41:21
[00:03:07]    |
[00:03:07] 41 |                sugg.substitution_parts.len() == 1 &&
[00:03:07]    |                     ^^^^^^^^^^^^^^^^^^
[00:03:07] 
```

Are you working on this @neosilky? Need any help?

**Comment by arielb1** [comment]
Feel free to ask for help on IRC (on #rust-internals or #rustc) - you can ping me (arielb1/arielby) personally.

**Comment by neosilky** [comment]
Sorry @arielb1, I'm just prepping for an exam tomorrow... I will fix this up after then!

**Comment by aidanhs** [comment]
Hi @neosilky, just want to make sure this hasn't dropped off your radar! Do you think you'll get time to look at it?

**Comment by Mark-Simulacrum** [comment]
Hm, that same test failure that @arielb1 noticed is still present, and I'm not actually quite sure why. @arielb1 is there a chance you could investigate and comment here?

**Comment by arielb1** [comment]
Taking a fresh look  at the test failures, which is:
```
[00:45:34] ---- [run-make] run-make/alloc-extern-crates stdout ----
[00:45:34] 	
[00:45:34] error: make failed
[00:45:34] status: exit code: 2
[00:45:34] command: "make"
[00:45:34] stdout:
[00:45:34] ------------------------------------------
[00:45:34] make[1]: Entering directory '/checkout/src/test/run-make/alloc-extern-crates'
[00:45:34] LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/alloc-extern-crates.stage2-x86_64-unknown-linux-gnu:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib:" '/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc' --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/alloc-extern-crates.stage2-x86_64-unknown-linux-gnu -L /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/alloc-extern-crates.stage2-x86_64-unknown-linux-gnu  fakealloc.rs
[00:45:34] LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/alloc-extern-crates.stage2-x86_64-unknown-linux-gnu:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib:" '/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc' --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/alloc-extern-crates.stage2-x86_64-unknown-linux-gnu -L /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/alloc-extern-crates.stage2-x86_64-unknown-linux-gnu  ../../../liballoc/lib.rs --cfg feature=\"external_crate\" --extern external=/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/alloc-extern-crates.stage2-x86_64-unknown-linux-gnu/libfakealloc.rlib
[00:45:34] Makefile:4: recipe for target 'all' failed
[00:45:34] make[1]: Leaving directory '/checkout/src/test/run-make/alloc-extern-crates'
[00:45:34] 
[00:45:34] ------------------------------------------
[00:45:34] stderr:
[00:45:34] ------------------------------------------
[00:45:34] make[1]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
[00:45:34] error: language item required, but not found: `panic_fmt`
[00:45:34] 
[00:45:34] error: language item required, but not found: `eh_personality`
[00:45:34] 
[00:45:34] error: aborting due to previous error(s)
[00:45:34] 
[00:45:34] make[1]: *** [all] Error 101
[00:45:34] 
[00:45:34] ------------------------------------------
[00:45:34] 
[00:45:34] thread '[run-make] run-make/alloc-extern-crates' panicked at 'explicit panic', /checkout/src/tools/compiletest/src/runtest.rs:2480
[00:45:34] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[00:45:34] 
[00:45:34] 
[00:45:34] failures:
[00:45:34]     [run-make] run-make/alloc-extern-crates
```

It really looks like the problem is that https://github.com/rust-lang/rust/blob/master/src/test/run-make/alloc-extern-crates/Makefile#L5 is missing a `--crate-type=rlib`:
```makefile
-include ../tools.mk

all:
	$(RUSTC) fakealloc.rs
	$(RUSTC) ../../../liballoc/lib.rs --cfg feature=\"external_crate\" --extern external=$(TMPDIR)/$(shell $(RUSTC) --print file-names fakealloc.rs)
        # ^ should be `--crate-type=rlib`
```

**Comment by arielb1** [comment]
After you fix the test, you can check it by running `./x.py test src/test/run-make`

**Comment by arielb1** [comment]
🕸 Closing PR to keep the queue clean 🕸. This looks like a simple fix - feel free to reopen with it.

**Pull request closed without merging**